atom/renderer/api/atom_api_renderer_ipc.cc

```diff
@@ -109,32 +110,50 @@ class IPCRenderer : public mate::Wrappable<IPCRenderer> {
   }
 
  private:
-  static void SendMessageSyncOnWorkerThread(
+  void SendMessageSyncOnWorkerThread(
       atom::mojom::ElectronBrowserPtrInfo* interface_info,
       base::WaitableEvent* event,
       base::Value* result,
       bool internal,
       const std::string& channel,
       const base::ListValue* arguments) {
     atom::mojom::ElectronBrowserPtr browser_ptr(std::move(*interface_info));
-    browser_ptr->MessageSync(
+    worker_thread_electron_browser_ptr_ = std::move(browser_ptr);
+
+    worker_thread_electron_browser_ptr_.set_connection_error_handler(
+        base::BindOnce(&IPCRenderer::HandleMojoConnectionErrorOnWorkerThread,
+                       base::Unretained(&worker_thread_electron_browser_ptr_),
+                       base::Unretained(interface_info),
+                       base::Unretained(event)));
+    worker_thread_electron_browser_ptr_->MessageSync(
         internal, channel, arguments->Clone(),
         base::BindOnce(&IPCRenderer::ReturnSyncResponseToMainThread,
-                       std::move(browser_ptr), base::Unretained(interface_info),
+                       base::Unretained(&worker_thread_electron_browser_ptr_),
+                       base::Unretained(interface_info),
                        base::Unretained(event), base::Unretained(result)));
   }
   static void ReturnSyncResponseToMainThread(
-      atom::mojom::ElectronBrowserPtr ptr,
+      atom::mojom::ElectronBrowserPtr* ptr,
       atom::mojom::ElectronBrowserPtrInfo* interface_info,
       base::WaitableEvent* event,
       base::Value* result,
       base::Value response) {
     *result = std::move(response);
-    *interface_info = ptr.PassInterface();
+    *interface_info = ptr->PassInterface();
+    event->Signal();
+  }
+  // If the other end of the message pipe disconnects, ensure we don't hang the
+  // main thread forever.
+  static void HandleMojoConnectionErrorOnWorkerThread(
+      atom::mojom::ElectronBrowserPtr* ptr,
+      atom::mojom::ElectronBrowserPtrInfo* interface_info,
+      base::WaitableEvent* event) {
+    *interface_info = ptr->PassInterface();
```

Comment:
We should probably record somewhere that an error occurred, so we can throw instead of returning an empty value.

Reply:
hm... what does it mean to return the interface once the connection is severed? perhaps we should leave the interface empty in this case? also we should make sure not to attempt to send more messages across the interface once the first failure has been detected.

Reply:
Intuitively, it makes sense to PassInterface — that's the unbind operation. Changing the question slightly but with similar intent: should one be allowed to re-bind after an error? This change represents no regression on that front, as now we're simply aware of errors, with no effect on how we or Mojo responds to them. I agree, it would be a good thing to consider, but it feels reasonably out of scope for this PR, especially since this change is targeted at stale 6-0-x.